File: python/graph_general/cource_schedule_II.py
from typing import List
from collections import deque, defaultdict


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        in_degree = [0] * numCourses

        for course, prereq in prerequisites:
            graph[prereq].append(course)

            in_degree[course] += 1


        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        order = []

        while queue:
            course = queue.popleft()

            order.append(course)

            for next_course in graph[course]:
                in_degree[next_course] -= 1

                if in_degree[next_course] == 0:
                    queue.append(next_course)


        if len(order) != numCourses:
            return []

        return order

File: golang/graph_general/cource_schedule_II.go
package main

func findOrder(numCourses int, prerequisites [][]int) []int {
	graph := make(map[int][]int)
	inDegree := make([]int, numCourses)

	for _, pre := range prerequisites {
		course, prereq := pre[0], pre[1]

		graph[prereq] = append(graph[prereq], course)

		inDegree[course]++
	}

	queue := []int{}

	for i := 0; i < numCourses; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	order := []int{}

	for len(queue) > 0 {
		course := queue[0]

		queue = queue[1:]

		order = append(order, course)

		for _, nextCourse := range graph[course] {
			inDegree[nextCourse]--

			if inDegree[nextCourse] == 0 {
				queue = append(queue, nextCourse)
			}
		}
	}

	if len(order) != numCourses {
		return []int{}
	}

	return order
}
